File: problems/002/002.cpp
#include<cstring>
#include<ctime>
#include<iostream>
#include<string>

using namespace std;

int solution() {
  int sum = 0, curr = 1, next = 2;

  while (curr < 4000000) {
    if (curr % 2 == 0)
      sum += curr;

    int temp = next;
    next += curr;
    curr = temp;
  }

  return sum;
}

int main(int argc, char* argv[]) {
  if (argc == 2 && strcmp(argv[1], "-a") == 0) {
    cout << solution() << endl;
    return 0;
  }

  for (string line; getline(cin, line);) {
    int iters = atoi(line.c_str());

    struct timespec start, end;
    clock_gettime(CLOCK_MONOTONIC_RAW, &start);
    for (int i = 0; i < iters; i++) {
      solution();
    }
    clock_gettime(CLOCK_MONOTONIC_RAW, &end);

    long int secs = end.tv_sec - start.tv_sec;
    long int nsecs = end.tv_nsec - start.tv_nsec;
    cout << secs * 1000000000 + nsecs << endl;
  }

  return 0;
}

File: problems/002/002.hs
import Control.Exception.Base (evaluate)
import System.Clock
import System.Environment (getArgs)
import System.IO

input :: Int
input = 4000000

fib :: Int -> Int
fib = fib' 1 2
  where fib' curr _ 0 = curr
        fib' curr next n = fib' next (curr+next) (n-1)

solution :: Int -> Int
solution n
  = sum
  . filter even
  . takeWhile (<n)
  . map fib
  $ [1..]

run :: (a -> b) -> a -> Int -> IO ()
run fun inp iters
  | iters <= 0 = return ()
  | otherwise = evaluate (fun inp) >> run fun inp (iters-1)

benchmark :: (a -> b) -> a -> Int -> IO Int
benchmark fun inp iters = do
  start <- getTime Monotonic
  run fun inp iters
  end <- getTime Monotonic
  let secs = sec end - sec start
  let nsecs = nsec end - nsec start
  let elapsed = 1000000000 * secs + nsecs
  return elapsed

benchmarkLoop :: IO ()
benchmarkLoop = do
  line <- getLine
  let iters = read line
  time <- benchmark solution input iters
  print time
  hFlush stdout
  benchmarkLoop

main = do
  args <- getArgs
  case args of
    [ "-a" ] -> print $ solution input
    [] -> benchmarkLoop
